Initial java openliberty stack

scottkurz opened the pull request:
### Checklist:

- [x] Read the [Code of Conduct](https://github.com/appsody/website/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md).

- [x] Followed the [commit message guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).

- [x] Stack adheres to [Appsody stack structure](https://github.com/appsody/website/blob/master/content/docs/stacks/stack-structure.md).


### Modifying an existing stack:

- N/A

<!--- Describe your changes in detail -->

### Contributing a new stack:

- Describe how application dependencies are managed:

This stack is Maven-based and this aspect functions largely like the java-microprofile predecessor.

The app pom.xml references the stack's pom.xml as its parent POM.    When launched, e.g. via `appsody run`, the container mounts the host .m2 repo to leverage the already-downloaded dependencies in the user's host environment.  

- Explain how Appsody file watcher is utilized:

We are currently only watching **pom.xml**, since lauching `mvn liberty:dev` launches a "watcher" with the liberty maven plugin (i.e. it launches in "dev mode").    (And this special case should probably be removed and folded into the others soon, using newer liberty-maven-plugin versions).

- Describe other Appsody environment variables defined by the stack image:

No special env vars.  It would help though to describe the way that maven profiles are used to differentiate between the "local dev" and "build/deploy" modes, as well as in the host-only (non-Docker) "mode" (which may not be required by Appsody but can still be useful).

- Describe any limitations and known issues:

See the [README.md](https://github.com/appsody/stacks/blob/master/incubator/java-openliberty/README.md) for more info (or [this link](https://github.com/scottkurz/stacks/blob/initial-java-openliberty-stack/incubator/java-openliberty/README.md) in the meantime until the PR is merged) especially the section on **Windows 10** issues.

### Related Issues:
Fixes #584, fixes #364

neeraj-laad commented:
Does this PR replace #552 ?

scottkurz commented:
> 
> 
> Does this PR replace #552 ?

Yes (closed that now).

yeekangc commented on incubator/java-openliberty/image/project/pom.xml:
```diff
@@ -0,0 +1,125 @@
+                                    <artifactId>${liberty.artifactId}</artifactId>
+                                    <version>${version.openliberty-runtime}</version>
+                                    <type>zip</type>
+                                </assemblyArtifact>
```
Do we need the assemblyArtifact?

scottkurz replied:
I think it's helpful to install this in the base, stack image so the user doesn't have to wait as long on `appsody run` to install it.

yeekangc replied:
Thanks, Scott.  We can update later to use the latest way to specify the runtime to install.

yeekangc commented:
@neeraj-laad, we are ready for a 0.1 release.  Can you or someone on the team merge?

scottkurz commented on incubator/java-openliberty/README.md:
```diff
@@ -0,0 +1,74 @@
+# Java OpenLiberty Stack
```
Good point.. made this change.

yeekangc commented:
I don't know who exactly has write access.  Tagging @kylegc @sympatheticmoose in case they can take a look and merge too.  Thank you.

scottkurz commented on incubator/java-openliberty/image/project/Dockerfile:
```diff
@@ -0,0 +1,35 @@
+# Step 2: Package Open Liberty image
+FROM open-liberty:kernel-java8-openj9
+
+COPY --chown=1001:0 --from=0 /config/ /config/
```
Ah..thanks.

scottkurz commented on incubator/java-openliberty/image/project/pom.xml:
```diff
@@ -0,0 +1,123 @@
+
+    <groupId>dev.appsody</groupId>
+    <artifactId>java-openliberty</artifactId>
+    <version>0.1.0</version>
```
Thanks...definitely want to do this soon, but was planning on using a later PR.

scottkurz commented:
Thx @Kamran64 for the review.. hopefully my comments are clear enough, but since I rebased let me just note since the last PR, I just added the one new commit https://github.com/appsody/stacks/pull/573/commits/881396e698331be3d6ed1af9e3e3a3760db3e1d6.

scottkurz commented on incubator/java-openliberty/image/config/app-deploy.yaml:
```diff
@@ -0,0 +1,30 @@
+    type: NodePort
+    port: APPSODY_PORT
+    annotations:
+      prometheus.io/scrape: 'true'
```
Thx for the suggestion.  Will give this some thought later on as we get more into metrics and monitoring.

scottkurz commented:
Fixes #584, fixes #364

skoh7645 commented:
When comparing the time taken to execute run/build commands on the this `java-openliberty` stack vs the `java-microprofile` stack
Open Liberty
1. First appsody run - 43s
2. Recompile run on change - 13s
3. First build - 4m 4s

Microprofile
1. First appsody run - 1m 33s
2. Recompile run on change - 26s
3. First build - 5m 15s

Kamran64 commented:
@scottkurz I can access all endpoints except for javametrics-dash. Are you seeing this behaviour also?

yeekangc commented:
Do we need javametrics still?  Codewind is moving away from that and will use standard metrics endpoints going forward?
